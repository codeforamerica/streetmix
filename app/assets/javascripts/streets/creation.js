var NEW_STREET_DEFAULT = 1;
var NEW_STREET_EMPTY = 2;

function _makeDefaultStreet() {
  ignoreStreetChanges = true;
  _prepareDefaultStreet();
  _setUpdateTimeToNow();

  _resizeStreetWidth();
  _updateStreetName();
  _createDomFromData();
  _segmentsChanged();
  _updateShareMenu();

  ignoreStreetChanges = false;
  lastStreet = _trimStreetData(street);

  _saveStreetToServer(false);
}

function _onNewStreetDefaultClick() {
  settings.newStreetPreference = NEW_STREET_DEFAULT;
  _saveSettingsLocally();

  _makeDefaultStreet();
}

function _onNewStreetEmptyClick() {
  settings.newStreetPreference = NEW_STREET_EMPTY;
  _saveSettingsLocally();

  ignoreStreetChanges = true;
  _prepareEmptyStreet();

  _resizeStreetWidth();
  _updateStreetName();
  _createDomFromData();
  _segmentsChanged();
  _updateShareMenu();

  ignoreStreetChanges = false;
  lastStreet = _trimStreetData(street);

  _saveStreetToServer(false);
}

function _onNewStreetLastClick() {
  _fetchLastStreet();
}
