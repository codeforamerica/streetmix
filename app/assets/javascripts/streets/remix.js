var remixOnFirstEdit = false;

// Auto “promote” (remix) the street if you just signed in and the street
// was anonymous
var promoteStreet = false;

function _remixStreet() {
  if (readOnly) {
    return;
  }

  remixOnFirstEdit = false;

  if (signedIn) {
    _setStreetCreator(signInData.userId, signInData.twitterId);
  } else {
    _setStreetCreator(null, null);
  }

  street.originalStreetId = street.id;
  street.editCount = 0;
  //console.log('editCount = 0 on remix!');

  _unifyUndoStack();

  if (undoStack[undoPosition - 1] && (undoStack[undoPosition - 1].name != street.name)) {
    // The street was remixed as a result of editing its name. Don’t be
    // a douche and add (remixed) to it then.
    var dontAddSuffix = true;
  } else {
    var dontAddSuffix = false;
  }

  if (!promoteStreet && !dontAddSuffix) {
    _addRemixSuffixToName();
  }

  var transmission = _packServerStreetData();

  _newBlockingAjaxRequest(msg('BLOCKING_REMIXING'),
      {
        // TODO const
        url: API_URL + 'v2/streets',
        type: 'POST',
        data: transmission,
        dataType: 'json',
        contentType: 'application/json',
        headers: { 'Authorization': _getAuthHeader() }
      }, _receiveRemixedStreet
  );
}

function _receiveRemixedStreet(data) {
  if (!promoteStreet) {
    if (signedIn) {
      _statusMessage.show(msg('STATUS_NOW_REMIXING'));
    } else {
      _statusMessage.show(msg('STATUS_NOW_REMIXING_SIGN_IN', { signInUrl: URL_SIGN_IN_REDIRECT }));
    }
  }

  _setStreetId(data.id, data.namespacedId);
  _updateStreetName();

  _saveStreetToServer(false);
}
