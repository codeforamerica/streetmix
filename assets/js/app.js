/*
 * Streetmix
 *
 * Front-end (mostly) by Marcin Wichary, Code for America fellow in 2013.
 *
 */

//= require vendor/modernizr
//= require vendor/moment
//= require vendor/jquery-2.1.0.min
//= require vendor/jquery.cookie

//= require_self
//= require_tree constants
//= require_tree util
//= require misc
//= require_tree app
//= require_tree segments
//= require_tree streets
//= require_tree users

var app = {};
