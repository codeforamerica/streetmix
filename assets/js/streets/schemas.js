var LATEST_SCHEMA_VERSION = 16;
// 1: starting point
// 2: adding leftBuildingHeight and rightBuildingHeight
// 3: adding leftBuildingVariant and rightBuildingVariant
// 4: adding transit shelter elevation
// 5: adding another lamp type (traditional)
// 6: colored streetcar lanes
// 7: colored bus and light rail lanes
// 8: colored bike lane
// 9: second car type: truck
// 10: sidewalk density
// 11: unify median and planting strip into divider
// 12: getting rid of small tree
// 13: bike rack elevation
// 14: wayfinding has three types
// 15: sidewalks have rand seed
// 16: stop saving undo stack

function _incrementSchemaVersion(street) {
  if (!street.schemaVersion) {
    street.schemaVersion = 1;
  }

  switch (street.schemaVersion) {
    case 1:
      street.leftBuildingHeight = DEFAULT_BUILDING_HEIGHT_LEFT;
      street.rightBuildingHeight = DEFAULT_BUILDING_HEIGHT_RIGHT;
      break;
    case 2:
      street.leftBuildingVariant = DEFAULT_BUILDING_VARIANT_LEFT;
      street.rightBuildingVariant = DEFAULT_BUILDING_VARIANT_RIGHT;
      break;
    case 3:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if (segment.type == 'transit-shelter') {
          var variant = _getVariantArray(segment.type, segment.variantString);
          variant['transit-shelter-elevation'] = 'street-level';
          segment.variantString = _getVariantString(variant);
        }
      }
      break;
    case 4:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if (segment.type == 'sidewalk-lamp') {
          var variant = _getVariantArray(segment.type, segment.variantString);
          variant['lamp-type'] = 'modern';
          segment.variantString = _getVariantString(variant);
        }
      }
      break;
    case 5:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if (segment.type == 'streetcar') {
          var variant = _getVariantArray(segment.type, segment.variantString);
          variant['public-transit-asphalt'] = 'regular';
          segment.variantString = _getVariantString(variant);
        }
      }
      break;
    case 6:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if ((segment.type == 'bus-lane') || (segment.type == 'light-rail')) {
          var variant = _getVariantArray(segment.type, segment.variantString);
          variant['public-transit-asphalt'] = 'regular';
          segment.variantString = _getVariantString(variant);
        }
      }
      break;
    case 7:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if (segment.type == 'bike-lane') {
          var variant = _getVariantArray(segment.type, segment.variantString);
          variant['bike-asphalt'] = 'regular';
          segment.variantString = _getVariantString(variant);
        }
      }
      break;
    case 8:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if (segment.type == 'drive-lane') {
          var variant = _getVariantArray(segment.type, segment.variantString);
          variant['car-type'] = 'car';
          segment.variantString = _getVariantString(variant);
        }
      }
      break;
    case 9:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if (segment.type == 'sidewalk') {
          var variant = _getVariantArray(segment.type, segment.variantString);
          variant['sidewalk-density'] = 'normal';
          segment.variantString = _getVariantString(variant);
        }
      }
      break;
    case 10:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if (segment.type == 'planting-strip') {
          segment.type = 'divider';

          if (segment.variantString == '') {
            segment.variantString = 'planting-strip';
          };
        } else if (segment.type == 'small-median') {
          segment.type = 'divider';
          segment.variantString = 'median';
        }
      }
      break;
    case 11:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if (segment.type == 'divider') {
          if (segment.variantString == 'small-tree') {
            segment.variantString = 'big-tree';
          };
        } else if (segment.type == 'sidewalk-tree') {
          if (segment.variantString == 'small') {
            segment.variantString = 'big';
          };
        }
      }
      break;
    case 12:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if (segment.type == 'sidewalk-bike-rack') {
          var variant = _getVariantArray(segment.type, segment.variantString);
          variant['bike-rack-elevation'] = 'sidewalk';
          segment.variantString =  _getVariantString(variant);
        }
      }
      break;
    case 13:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if (segment.type == 'sidewalk-wayfinding') {
          var variant = _getVariantArray(segment.type, segment.variantString);
          variant['wayfinding-type'] = 'large';
          segment.variantString =  _getVariantString(variant);
        }
      }
      break;
    case 14:
      for (var i in street.segments) {
        var segment = street.segments[i];
        if (segment.type == 'sidewalk') {
          segment.randSeed = 35;
        }
      }
      break;
    case 15:
      undoStack = [];
      undoPosition = 0;
      break;
  }

  street.schemaVersion++;
}

function _updateToLatestSchemaVersion(street) {
  var updated = false;
  while (!street.schemaVersion || (street.schemaVersion < LATEST_SCHEMA_VERSION)) {
    _incrementSchemaVersion(street);
    updated = true;
  }

  return updated;
}
