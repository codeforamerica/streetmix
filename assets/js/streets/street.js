var street = {
  schemaVersion: LATEST_SCHEMA_VERSION,

  id: null,
  creatorId: null,
  namespacedId: null,
  originalStreetId: null, // id of the street the current street is remixed from (could be null)
  name: null,
  editCount: null, // Since new street or remix · FIXME: can be null (meaning = don’t touch), but this will change

  width: null,
  occupiedWidth: null, // Can be recreated, do not save
  remainingWidth: null, // Can be recreated, do not save

  leftBuildingHeight: null,
  rightBuildingHeight: null,
  leftBuildingVariant: null,
  rightBuildingVariant: null,

  segments: [],

  units: null
};
