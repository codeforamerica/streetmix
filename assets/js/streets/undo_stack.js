var UNDO_LIMIT = 1000

var FLAG_SAVE_UNDO = false; // true to save undo with street data, false to not save undo

var undoStack = []
var undoPosition = 0
var ignoreStreetChanges = false

function _undoRedo (undo) {
  if (undo && !_isUndoAvailable()) {
    showStatusMessage(msg('STATUS_NOTHING_TO_UNDO'))
  } else if (!undo && !_isRedoAvailable()) {
    showStatusMessage(msg('STATUS_NOTHING_TO_REDO'))
  } else {
    if (undo) {
      undoStack[undoPosition] = _trimStreetData(street)
      undoPosition--
    } else {
      undoPosition++
    }
    street = _.cloneDeep(undoStack[undoPosition])
    _setUpdateTimeToNow()

    _infoBubble.hide()
    _infoBubble.hideSegment()
    _infoBubble.dontConsiderShowing()

    _updateEverything(true)
    hideStatusMessage()
  }

  trackEvent('INTERACTION', 'UNDO', null, null, true)
}

function _clearUndoStack () {
  undoStack = []
  undoPosition = 0
  _updateUndoButtons()
}

function _undo () {
  _undoRedo(true)
}

function _redo () {
  _undoRedo(false)
}

function _trimUndoStack () {
  // TODO optimize
  while (undoPosition >= UNDO_LIMIT) {
    undoPosition--
    undoStack = undoStack.slice(1)
  }
}

function _createNewUndo () {
  // This removes future undo path in case we undo a few times and then do
  // something undoable.
  undoStack = undoStack.splice(0, undoPosition)
  undoStack[undoPosition] = _.cloneDeep(lastStreet)
  undoPosition++

  _trimUndoStack()
  _unifyUndoStack()
}

function _createNewUndoIfNecessary (lastStreet, currentStreet) {
  if (lastStreet.name != currentStreet.name) {
    return
  }

  _createNewUndo()
}

function _isUndoAvailable () {
  // Don’t allow undo/redo unless you own the street

  return (undoPosition > 0) && !remixOnFirstEdit
}

function _isRedoAvailable () {
  // Don’t allow undo/redo unless you own the street

  return (undoPosition < undoStack.length - 1) && !remixOnFirstEdit
}

function _updateUndoButtons () {
  document.querySelector('#undo').disabled = !_isUndoAvailable()
  document.querySelector('#redo').disabled = !_isRedoAvailable()
}

function _unifyUndoStack () {
  for (var i = 0; i < undoStack.length; i++) {
    undoStack[i].id = street.id
    undoStack[i].name = street.name
    undoStack[i].namespacedId = street.namespacedId
    undoStack[i].creatorId = street.creatorId
    undoStack[i].updatedAt = street.updatedAt
  }
}
