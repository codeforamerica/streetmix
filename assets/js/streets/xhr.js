function _createNewStreetOnServer() {
  if (settings.newStreetPreference == NEW_STREET_EMPTY) {
    _prepareEmptyStreet();
  } else {
    _prepareDefaultStreet();
  }

  var transmission = _packServerStreetData();

  $.ajax({
    // TODO const
    url: API_URL + 'v1/streets',
    data: transmission,
    type: 'POST',
    dataType: 'json',
    headers: { 'Authorization': _getAuthHeader() }
  }).done(_receiveNewStreet)
  .fail(_errorReceiveNewStreet);
}

function _receiveNewStreet(data) {
  _setStreetId(data.id, data.namespacedId);

  _saveStreetToServer(true);
}

function _errorReceiveNewStreet(data) {
  _showError(ERRORS.NEW_STREET_SERVER_FAILURE, true);
}

function _getFetchStreetUrl() {
  // TODO const
  if (street.creatorId) {
    var url = API_URL + 'v1/streets?namespacedId=' +
        encodeURIComponent(street.namespacedId) + '&creatorId=' +
        encodeURIComponent(street.creatorId);
  } else {
    var url = API_URL + 'v1/streets?namespacedId=' +
        encodeURIComponent(street.namespacedId);
  }

  return url;
}

function _fetchStreetFromServer() {
  var url = _getFetchStreetUrl();

  $.ajax({
    url: url,
    dataType: 'json',
    type: 'GET'
  }).done(_receiveStreet).fail(_errorReceiveStreet);
}

function _errorReceiveStreet(data) {
  if ((mode == MODES.CONTINUE) || (mode == MODES.USER_GALLERY) ||
      (mode == MODES.ABOUT) || (mode == MODES.GLOBAL_GALLERY)) {
    _goNewStreet();
  } else {
    if ((data.status == 404) || (data.status == 410)) {
      if (street.creatorId) {
        if (data.status == 410) {
          mode = MODES.STREET_410_BUT_LINK_TO_USER;
        } else {
          mode = MODES.STREET_404_BUT_LINK_TO_USER;
        }
      } else {
        mode = MODES.STREET_404;
      }
      // TODO swap for showError (here and elsewhere)
      _processMode();
    } else {
      _showError(ERRORS.NEW_STREET_SERVER_FAILURE, true);
    }
  }
}

function _saveStreetToServer(initial) {
  if (readOnly) {
    return;
  }

  var transmission = _packServerStreetData();

  if (initial) {
    // blocking

    $.ajax({
      // TODO const
      url: API_URL + 'v1/streets/' + street.id,
      data: transmission,
      dataType: 'json',
      type: 'PUT',
      contentType: 'application/json',
      headers: { 'Authorization': _getAuthHeader() }
    }).done(_confirmSaveStreetToServerInitial);
  } else {
    _newNonblockingAjaxRequest({
      // TODO const
      url: API_URL + 'v1/streets/' + street.id,
      data: transmission,
      dataType: 'json',
      type: 'PUT',
      contentType: 'application/json',
      headers: { 'Authorization': _getAuthHeader() }
    }, false);
  }
}

function _clearScheduledSavingStreetToServer() {
  window.clearTimeout(saveStreetTimerId);
}

function _fetchStreetForVerification() {
  // Donâ€™t do it with any network services pending
  if (_getNonblockingAjaxRequestCount() || blockingAjaxRequestInProgress ||
      saveStreetIncomplete || abortEverything || remixOnFirstEdit) {
    return;
  }

  var url = _getFetchStreetUrl();

  latestRequestId = _getUniqueRequestHeader();
  latestVerificationStreet = _trimStreetData(street);

  $.ajax({
    url: url,
    dataType: 'json',
    type: 'GET',
    // TODO const
    headers: { 'X-Streetmix-Request-Id': latestRequestId }
  }).done(_receiveStreetForVerification).fail(_errorReceiveStreetForVerification);
}

function _receiveStreetForVerification(transmission, textStatus, request) {
  var requestId = parseInt(request.getResponseHeader('X-Streetmix-Request-Id'));

  if (requestId != latestRequestId) {
    return;
  }

  var localStreetData = _trimStreetData(latestVerificationStreet);
  var serverStreetData = _trimStreetData(_unpackStreetDataFromServerTransmission(transmission));

  if (JSON.stringify(localStreetData) != JSON.stringify(serverStreetData)) {
    console.log('NOT EQUAL');
    console.log('-');
    console.log(JSON.stringify(localStreetData));
    console.log('-');
    console.log(JSON.stringify(serverStreetData));
    console.log('-');
    console.log(transmission);

    _statusMessage.show(msg('STATUS_RELOADED_FROM_SERVER'));

    _infoBubble.suppress();

    _unpackServerStreetData(transmission, null, null, false);
    _updateEverything(true);

    _eventTracking.track(TRACK_CATEGORY_EVENT,
        TRACK_ACTION_STREET_MODIFIED_ELSEWHERE, null, null, false);
  }
}

function _errorReceiveStreetForVerification(data) {
  // 404 should never happen here, since 410 designates streets that have
  // been deleted (but remain hidden on the server)

  if (signedIn && ((data.status == 404) || (data.status == 410))) {
    _showError(ERRORS.STREET_DELETED_ELSEWHERE, true);
  }
}

function _receiveStreet(transmission) {
  _unpackServerStreetData(transmission, null, null, true);

  _propagateUnits();

  // TODO this is stupid, only here to fill some structures
  _createDomFromData();
  _createDataFromDom();

  serverContacted = true;
  _checkIfEverythingIsLoaded();
}

function _unpackStreetDataFromServerTransmission(transmission) {
  var street = _clone(transmission.data.street);

  street.creatorId = (transmission.creator && transmission.creator.id) || null;
  street.originalStreetId = transmission.originalStreetId || null;
  street.updatedAt = transmission.updatedAt || null;
  street.name = transmission.name || DEFAULT_NAME;

  // FIXME just read it and do 0 otherwise
  if (typeof transmission.data.street.editCount == 'undefined') {
    //console.log('editCount read is empty');
    street.editCount = null;
  } else {
    street.editCount = transmission.data.street.editCount;
    //console.log('editCount read is', street.editCount);
  }

  return street;
}

function _unpackServerStreetData(transmission, id, namespacedId, checkIfNeedsToBeRemixed) {
  street = _unpackStreetDataFromServerTransmission(transmission);

  if (transmission.data.undoStack) {
    undoStack = _clone(transmission.data.undoStack);
    undoPosition = transmission.data.undoPosition;
  } else {
    undoStack = [];
    undoPosition = 0;
  }

  var updatedSchema = _updateToLatestSchemaVersion(street);
  for (var i = 0; i < undoStack.length; i++) {
    if (_updateToLatestSchemaVersion(undoStack[i])) {
      updatedSchema = true;
    }
  }

  if (id) {
    _setStreetId(id, namespacedId);
  } else {
    _setStreetId(transmission.id, transmission.namespacedId);
  }

  if (checkIfNeedsToBeRemixed) {
    if (!signedIn || (street.creatorId != signInData.userId)) {
      remixOnFirstEdit = true;
    } else {
      remixOnFirstEdit = false;
    }

    if (updatedSchema && !remixOnFirstEdit) {
      _saveStreetToServer();
    }
  }
}

function _packServerStreetData() {
  var data = {};

  data.street = _trimStreetData(street);

  // Those go above data in the structure, so they need to be cleared here
  delete data.street.name;
  delete data.street.originalStreetId;
  delete data.street.updatedAt;

  // This will be implied through authorization header
  delete data.street.creatorId;

  if (FLAG_SAVE_UNDO) {
    data.undoStack = _clone(undoStack);
    data.undoPosition = undoPosition;
  }

  var transmission = {
    name: street.name,
    originalStreetId: street.originalStreetId,
    data: data
  }

  return JSON.stringify(transmission);
}

function _setStreetId(newId, newNamespacedId) {
  street.id = newId;
  street.namespacedId = newNamespacedId;

  _unifyUndoStack();

  _updateLastStreetInfo();
}

function _updateLastStreetInfo() {
  settings.lastStreetId = street.id;
  settings.lastStreetNamespacedId = street.namespacedId;
  settings.lastStreetCreatorId = street.creatorId;

  _saveSettingsLocally();
}
