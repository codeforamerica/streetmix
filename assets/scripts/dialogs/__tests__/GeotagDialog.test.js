/* eslint-env jest */
import React from 'react'
import { GeotagDialogWithIntl as GeotagDialog } from '../GeotagDialog'
import { shallowWithIntl as shallow } from '../../../../test/helpers/intl-enzyme-test-helper.js'
import { getRemixOnFirstEdit } from '../../streets/remix'

// Mock dependencies that could break tests
jest.mock('../../streets/remix', () => ({
  getRemixOnFirstEdit: jest.fn()
}))

function getTestComponent (addressInformation, street) {
  const testMarker = { lat: 0, lng: 0 }
  const testAddressInfo = addressInformation || {
    street: 'foo',
    id: 'foo'
  }
  const testStreet = street || {
    creatorId: 'foo',
    location: {
      label: 'foo',
      wofId: 'foo'
    }
  }

  return (
    <GeotagDialog
      street={testStreet}
      addressInformation={testAddressInfo}
      markerLocation={testMarker}
    />
  )
}

function updateProps (wrapper) {
  wrapper.setProps({
    markerLocation: { lat: 10, lng: 10 },
    addressInformation: {
      street: 'bar',
      id: 'bar'
    }
  })
}

describe('GeotagDialog', () => {
  it('renders without crashing', () => {
    const wrapper = shallow(
      <GeotagDialog
        street={{}}
        addressInformation={{}}
      />
    )
    expect(wrapper.exists()).toEqual(true)
  })

  it('does not allow a location to be confirmed when geocoded data does not have street data', () => {
    const testAddressInfo = {
      street: null
    }
    const wrapper = shallow(getTestComponent(testAddressInfo, {}))
    expect(wrapper.find('button')).toHaveLength(0)
  })

  it('allows a location to be confirmed when the current signed-in user is the street owner', () => {
    const wrapper = shallow(getTestComponent())
    getRemixOnFirstEdit.mockReturnValueOnce(false)
    updateProps(wrapper)
    expect(wrapper.instance().canEditLocation()).toEqual(true)
  })

  it('allows a location to be confirmed when the current anonymous user started this street', () => {
    const wrapper = shallow(getTestComponent())
    getRemixOnFirstEdit.mockReturnValueOnce(false)
    updateProps(wrapper)
    expect(wrapper.instance().canEditLocation()).toEqual(true)
  })

  it('does not allow a location to be confirmed when the current signed-in user is not the street owner', () => {
    const wrapper = shallow(getTestComponent())
    getRemixOnFirstEdit.mockReturnValueOnce(true)
    updateProps(wrapper)
    expect(wrapper.find('button')).toHaveLength(0)
  })

  it('does not allow a location to be confirmed when the current anonymous user is not the street owner and there is already an existing location attached', () => {
    const wrapper = shallow(getTestComponent())
    getRemixOnFirstEdit.mockReturnValueOnce(true)
    updateProps(wrapper)
    expect(wrapper.find('button')).toHaveLength(0)
  })

  it('allows a location to be confirmed when the current anonymous user is not the street owner but there is no existing location attached', () => {
    const testStreet = {
      creatorId: 'foo',
      location: null
    }
    const wrapper = shallow(getTestComponent(null, testStreet))
    getRemixOnFirstEdit.mockReturnValueOnce(true)
    updateProps(wrapper)
    expect(wrapper.instance().canEditLocation()).toEqual(true)
  })

  // Skip tests until we figure out how to actually test components that use render propsq
  it.skip('does not show error banner if geocoding services are available', () => {
    const wrapper = shallow(getTestComponent())
    wrapper.setState({ geocodeAvailable: true })
    expect(wrapper.find('.geotag-error-banner')).toHaveLength(0)
    expect(wrapper.find('.geotag-input-container')).toHaveLength(1)
  })

  it.skip('cripples dialog behavior if geocoding services are unavailable', () => {
    const wrapper = shallow(getTestComponent())
    wrapper.setState({ geocodeAvailable: false })
    expect(wrapper.find('.geotag-error-banner')).toHaveLength(1)
    expect(wrapper.find('.geotag-input-container')).toHaveLength(0)
  })
})
