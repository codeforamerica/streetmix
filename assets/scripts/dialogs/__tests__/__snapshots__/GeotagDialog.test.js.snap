// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeotagDialog renders 1`] = `
<DocumentFragment>
  <div
    class="dialog-box-container"
  >
    <div
      class="dialog-box-backdrop"
    />
    <div
      class="dialog-box"
      role="dialog"
    >
      <button
        class="close"
        title="Dismiss"
      >
        <svg
          class="svg-inline--fa fa-times"
        />
      </button>
      <div
        class="geotag-dialog"
      >
        <div
          class="geotag-error-banner"
        >
          Geocoding services are currently unavailable. You can view the map, but you won’t be able to change this street’s location.
        </div>
        <div
          class="leaflet-container leaflet-grab leaflet-touch-drag"
          style="position: relative;"
          tabindex="0"
        >
          <div
            class="leaflet-pane leaflet-map-pane"
            style="left: 0px; top: 0px;"
          >
            <div
              class="leaflet-pane leaflet-tile-pane"
            />
            <div
              class="leaflet-pane leaflet-shadow-pane"
            />
            <div
              class="leaflet-pane leaflet-overlay-pane"
            />
            <div
              class="leaflet-pane leaflet-marker-pane"
            />
            <div
              class="leaflet-pane leaflet-tooltip-pane"
            />
            <div
              class="leaflet-pane leaflet-popup-pane"
            />
          </div>
          <div
            class="leaflet-control-container"
          >
            <div
              class="leaflet-top leaflet-left"
            >
              <div
                class="leaflet-control-zoom leaflet-bar leaflet-control"
              >
                <a
                  aria-label="Zoom in"
                  class="leaflet-control-zoom-in"
                  href="#"
                  role="button"
                  title="Zoom in"
                >
                  +
                </a>
                <a
                  aria-label="Zoom out"
                  class="leaflet-control-zoom-out"
                  href="#"
                  role="button"
                  title="Zoom out"
                >
                  −
                </a>
              </div>
            </div>
            <div
              class="leaflet-top leaflet-right"
            />
            <div
              class="leaflet-bottom leaflet-left"
            />
            <div
              class="leaflet-bottom leaflet-right"
            >
              <div
                class="leaflet-control-attribution leaflet-control"
              >
                <a
                  href="https://leafletjs.com"
                  title="A JS library for interactive maps"
                >
                  Leaflet
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
